torch/csrc/distributed/c10d/symm_mem/CUDASymmetricMemory.cu

```diff
@@ -543,28 +581,46 @@ static void init_multicast_for_block(
           << "\". Gracefully skipping multicast initialization. "
           << "However, this is unexpected. Please report the issue on GitHub.";
       // Allow peers gracefully skip multicast initialization by sending -1
-      ipc_channel.broadcast_fds(rank, 0, pids, -1);
+      // TODO: allow graceful skip for fabric
+      if constexpr (!use_fabric_handle) {
+        ipc_channel.broadcast_fds(rank, 0, pids, -1);
+      }
       return;
     }
 
-    int mc_fd;
-    // using the CUDA Driver API to export a multicast object into a POSIX file descriptor.
+    McHandleType mc_exported_handle;
+    // using the CUDA Driver API to export a multicast object into a POSIX file
+    // descriptor.
     C10_CUDA_DRIVER_CHECK(driver_api->cuMemExportToShareableHandle_(
-        &mc_fd, mc_handle, CU_MEM_HANDLE_TYPE_POSIX_FILE_DESCRIPTOR, 0));
-    ipc_channel.broadcast_fds(rank, 0, pids, mc_fd);
-    // Ref count is incremented as soon as SCM_RIGHTS send happens
-    close(mc_fd);
+        &mc_exported_handle, mc_handle, handleType, 0));
+    if constexpr (!use_fabric_handle) {
+      ipc_channel.broadcast_fds(rank, 0, pids, mc_exported_handle);
+      // Ref count is incremented as soon as SCM_RIGHTS send happens
+      close(mc_exported_handle);
+    } else {
+      // TODO implement storeExchange.broadcast
+      storeExchange.all_gather(store, rank, world_size, mc_exported_handle);
+    }
```

Comment:
Is there chance to consolidate the two exchanges in the future? I guess `storeExchange.broadcast` can be also used for broadcasting fd ?

Reply:
storeExchange.broadcast just broadcasts bytes, for fds you need to send them through sockets. I guess it's possible to make a dummy ipc channel for fabric that just calls storeExchange.allgather or broadcast to simply broadcast fds.